File: homework/Truong_Duong/lesson_1/78_subsets.py
# Solution for https://leetcode.com/problems/subsets/

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def getBit(x, pos):
            return (x >> pos) & 1

        N = len(nums)
        ans = []

        for x in range(pow(2, N)):
            subset = []
            for i in range(N):
                if(getBit(x, i) == 1):
                    subset.append(nums[i])
            ans.append(subset)

        return ans

File: homework/Thu_Pham/lesson_1/78-subsets.js
//Solution for https://leetcode.com/problems/subsets/

/*
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
  let result = [];
  for (let i = 0; i < Math.pow(2, nums.length); i++) {
    let subset = [];
    for (let position = 0; position < nums.length; position++) {
      if (getBit(i, position) === 1) subset.push(nums[position]);
    }
    result.push(subset);
  }
  return result;
};

var getBit = function (number, position) {
  //O(1)
  return (number >> position) & 1;
};

//runtime O(2^n * n * 1), where n is the length of nums
//space O(2^n), where n is the length of nums
